app/__init__.py

```diff
@@ -146,13 +147,41 @@ def auth():
         # auth processors always assume a valid Authorization in header, if any
         headers = auth.process(request, headers)
     except jwt.ExpiredSignatureError:
-        current_app.logger.warning("Error while authenticating request", exc_info=True)
-        return Response(json.dumps({"error": "token_expired"}), status=401)
+        current_app.logger.warning(
+            f"Error while authenticating request, token expired. Target: {auth_arg}",
+            exc_info=True,
+        )
+        message = {
+            "error": "authentication",
+            "message": "token expired",
+            "target": auth_arg,
+        }
+        return Response(json.dumps(message), status=401)
+    except AttributeError as error:
+        if "access_token" in str(error):
```

Comment:
How do you know that an attribute error with this message stands for a problem in refreshing the token? Is this always the case?

Reply:
I did quite a few trials and the error I get when it's not possible to refresh the token is always a missing `access_token`. Iirc it happens in `web.get_valid_token` when trying to create the object `{"access_token": keycloak_oidc_client.access_token}` (the error is `AttributeError("'NoneType' object has no attribute 'access_token'")`).

Not sure if this is how that function is intended to work, maybe changing the code there to return something else (or throw a specific error) could be a cleaner solution.

Reply:
Yeah, that will work because [here](https://github.com/SwissDataScienceCenter/renku-gateway/blob/a9d2657c06f0e6b847f47f5946935628b0002634/app/auth/oauth_redis.py#L97) we just return `None` if the token can not be refreshed (which ist not very nice), hence no `access_token` attribute.

I think the cleanest would be to define a custom `token_refresh_error` as part of the `oauth_client` module and throw it already [here](https://github.com/SwissDataScienceCenter/renku-gateway/blob/a9d2657c06f0e6b847f47f5946935628b0002634/app/auth/oauth_client.py#L92) when the refreshing fails. Then we could catch that error [here](https://github.com/SwissDataScienceCenter/renku-gateway/blob/a9d2657c06f0e6b847f47f5946935628b0002634/app/auth/oauth_redis.py#L91), log the warning, re-raise the error and catch it again at the same point where you catch the attribute error now. Shouldn't be too much change, but we can also put this in a separate issue if you prefer to not tackle this now.